### PR TITLE
WIP: Dmr/unblock catchup

### LINKED ISSUE #15220: Federation catchup for a destination can block indefinitely on un-partial stating
**What we think happened**
1. A user faster joined a remote room
2. Some events were sent into the room, but not transmitted successfully
3. The user left the room before it became full state
4. The federation catchup loop tried to transmit events in the room, but blocked on full state. This also blocked \*all\* catchup transmission for the destination.

**Stack trace**
```
  File "/home/synapse/src/synapse/metrics/background_process_metrics.py", line 242, in run
    return await func(*args, **kwargs)
  File "/home/synapse/src/synapse/federation/sender/per_destination_queue.py", line 321, in _transaction_transmission_loop
    await self._catch_up_transmission_loop()
  File "/home/synapse/src/synapse/federation/sender/per_destination_queue.py", line 544, in _catch_up_transmission_loop
    new_pdus = await filter_events_for_server(
  File "/home/synapse/src/synapse/visibility.py", line 650, in filter_events_for_server
    event_to_memberships = await _event_to_memberships(
  File "/home/synapse/src/synapse/visibility.py", line 746, in _event_to_memberships
    event_to_state_ids = await storage.state.get_state_ids_for_events(
  File "/home/synapse/src/synapse/logging/opentracing.py", line 914, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/home/synapse/src/synapse/logging/opentracing.py", line 914, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/home/synapse/src/synapse/storage/controllers/state.py", line 267, in get_state_ids_for_events
    event_to_groups = await self.get_state_group_for_events(
  File "/home/synapse/src/synapse/logging/opentracing.py", line 914, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/home/synapse/src/synapse/logging/opentracing.py", line 914, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/home/synapse/src/synapse/storage/controllers/state.py", line 374, in get_state_group_for_events
    await self._partial_state_events_tracker.await_full_state(event_ids)
  File "/home/synapse/src/synapse/logging/opentracing.py", line 914, in _wrapper
    return await func(*args, **kwargs)  # type: ignore[misc]
  File "/home/synapse/src/synapse/storage/util/partial_state_events_tracker.py", line 82, in await_full_state
    logger.info(
```

### PR DESCRIPTION
**Pull request by DMRobertson**
Intended to fix #15220.

Follows from #14404.

Will be commitwise reviewable.

### CODE REVIEW

File: tests/federation/test_federation_catch_up.py [hunk 1]
```diff
@@ -500,3 +501,77 @@ def test_not_latest_event(self) -> None:
...
+            self.hs.get_storage_controllers().state._partial_state_events_tracker,
+            "await_full_state",
+            mock_await_full_state,
+        ):
+            self.get_success(per_dest_queue._catch_up_transmission_loop())
+
+        # ASSERT
+        # We should have:
+        # - not sent event 3: it's not ours, and the room is partial stated
+        # - fallen back to sending event 2: it's the most recent event in the room
```
Comment (DMRobertson) [inline tests/federation/test_federation_catch_up.py:572]: I think this is a little dodgy. In this situation, `host2` is in the room prior to `event_2` being created, so it's correct that we send `event_2` as the catchup event.

BUT: suppose that the order of event was

- event 2
- host2 joins
- event 3

then host 2 wouldn't be privvy to event 2. I think we'll either 
- block awaitng event2's full state, or
- incorrectly send event 2 to host2

I need to probably make another test case and stare at this some more.

File: synapse/visibility.py [hunk 2]
```diff
@@ -565,29 +575,40 @@ async def filter_events_for_server(
...
     see them.
 
+    For a fully stated room, the target server is allowed to see an event E if:
+      - the state at E has world readable or shared history vis, OR
+      - the state at E says that the target server is in the room.
+
+    For a partially stated room, the target server is allowed to see E if:
+      - E was created by this homeserver, AND:
+          - the partial state at E has world readable or shared history vis, OR
+          - the partial state at E says that the target server is in the room.
```
Comment (DMRobertson) [inline synapse/visibility.py:592]: The fully stated section is my reading of the original code.

The partially stated section is what I _think_ we are trying to do in the original code, and what I think we should be doing.

File: synapse/visibility.py [hunk 3]
```diff
@@ -587,7 +599,7 @@ async def filter_events_for_server(
         The filtered events.
     """
 
-    def is_sender_erased(event: EventBase, erased_senders: Dict[str, bool]) -> bool:
+    def is_sender_erased(event: EventBase, erased_senders: Mapping[str, bool]) -> bool:
```
Comment (DMRobertson) [inline synapse/visibility.py:611]: Prefer to take immutable types as arguments where possible---it makes the reader's life much easier!

File: synapse/visibility.py [hunk 4]
```diff
@@ -643,14 +643,13 @@ def check_event_is_visible(
     # this check but would base the filtering on an outdated view of the membership events.
 
     partial_state_invisible_events = set()
-    if filter_out_erased_senders:
```
Comment (DMRobertson) [inline synapse/visibility.py]: Am I missing something here?

Comment (MatMaul) [reply synapse/visibility.py]: I have put some [comments](https://github.com/matrix-org/synapse/commit/1526ff389f02d14d0df729bd6ea35836e758c449#r103679644) regarding that in the original PR.
We perhaps want to use a dedicated boolean like `filter_partial_state` for this (and keep your rename to `filter_out_erased_senders` for the rest of the code).
I think the logic I had in mind is that we don't care about having non local events when checking if we should use an event as an extremity for backfill: the remote server would then filter on his side anyway what need to be filtered, with a (potentially) a more current view than the joining server.

**Comment by MatMaul** [comment]
For people coming here after me, `Avoid state lookups for events we'll ignore` is the important commit.

File: synapse/visibility.py [hunk 5]
```diff
@@ -631,44 +652,49 @@ def check_event_is_visible(
     # otherwise a room could be fully joined after we retrieve those, which would then bypass
     # this check but would base the filtering on an outdated view of the membership events.
 
-    partial_state_invisible_events = set()
-    if not check_history_visibility_only:
```
Comment (MatMaul) [inline synapse/visibility.py:635]: According to some failing tests we need to keep a way to shortcut this code for the backfill use case.
In this case this function is called with backward extremities we want to use as a base for backfilling to `check each extremity in turn and ignore those which users on our server wouldn't be able to see.`
So there is no way here that we will leak events we shouldn't, so skipping this should be fine.
Why does it make the tests fail ? I haven't looked deep enough yet to understand.

**Comment by DMRobertson** [comment]
> I'm going to spin out the refactor parts of this in #15240.

And the remaining part is #15241.

**Pull request closed without merging**